src/core/net/ip6.cpp

```diff
@@ -934,29 +936,58 @@ otError Ip6::HandleExtensionHeaders(Message &    aMessage,
     return error;
 }
 
-otError Ip6::HandlePayload(Message &aMessage, MessageInfo &aMessageInfo, uint8_t aIpProto)
+void Ip6::HandlePayload(Message &          aMessage,
+                        MessageInfo &      aMessageInfo,
+                        uint8_t            aIpProto,
+                        Message::Ownership aMessageOwnership)
 {
-    otError error = OT_ERROR_NONE;
+    otError  error = OT_ERROR_NONE;
+    Message *message;
+
```

Comment:
We use the `message` only if the `aIpProto` is UDP or ICMP6. 
Can we check this at the top of this method, before potentially creating a `Clone()` of message (which may not then be used)?

Reply:
Good point! I was considering refactoring this part in a separate PR. Mainly it will carry the message lifecycle in its own, thus each layer can better handle it.